[SwiftUI] Introduce the NavigationDeciding protocol

rr-codes opened the pull request:
#### 7d88a900b0d51f0d827e99ffa15d6b729d0446b7
<pre>
[SwiftUI] Introduce the NavigationDeciding protocol
<a href="https://bugs.webkit.org/show_bug.cgi?id=283805">https://bugs.webkit.org/show_bug.cgi?id=283805</a>
<a href="https://rdar.apple.com/140672055">rdar://140672055</a>

Reviewed by Wenson Hsieh.

This protocol is designed to provide the subset of functionality of WKNavigationDelegate for making navigation
decisions and responding to authentication challenges.

The exception is `webView(_:authenticationChallenge:shouldAllowDeprecatedTLS:)`, which is intentionally excluded
as its functionality is effectively deprecated.

* Source/WebKit/UIProcess/API/Swift/WKNavigationDelegateAdapter.swift:
(WKNavigationDelegateAdapter.webView(_:decidePolicyFor:preferences:WKWebpagePreferences:)):
(WKNavigationDelegateAdapter.webView(_:decidePolicyFor:)):
(WKNavigationDelegateAdapter.webView(_:respondTo:URLCredential:)):
* Source/WebKit/UIProcess/API/Swift/WebPage+FrameInfo.swift: Added.
(FrameInfo.isMainFrame):
(FrameInfo.request):
(FrameInfo.securityOrigin):
(FrameInfo.wrapped):
* Source/WebKit/UIProcess/API/Swift/WebPage+NavigationDeciding.swift: Added.
(NavigationAction.source):
(NavigationAction.target):
(NavigationAction.navigationType):
(NavigationAction.request):
(NavigationAction.shouldPerformDownload):
(NavigationAction.buttonNumber):
(NavigationAction.modifierFlags):
(NavigationAction.wrapped):
(NavigationResponse.isForMainFrame):
(NavigationResponse.response):
(NavigationResponse.canShowMimeType):
(NavigationResponse.wrapped):
(NavigationDeciding.decidePolicy(for:preferences:)):
(NavigationDeciding.decidePolicy(for:)):
(NavigationDeciding.decideAuthenticationChallengeDisposition(for:URLCredential:)):
* Source/WebKit/UIProcess/API/Swift/WebPage+NavigationPreferences.swift: Added.
(preferredContentMode):
(allowsContentJavaScript):
(preferredHTTPSNavigationPolicy):
(isLockdownModeEnabled):
* Source/WebKit/UIProcess/API/Swift/WebPage.swift:
(WebPage_v0.callAsyncJavaScript(_:arguments:in:contentWorld:)):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKit Swift/Bundle+Extras.swift: Added.
(Bundle.testResources):
* Tools/TestWebKitAPI/Tests/WebKit Swift/WKWebViewSwiftOverlay.swift:
* Tools/TestWebKitAPI/Tests/WebKit Swift/WebPage.swift:
(TestNavigationDecider.preferencesMutation):
(TestNavigationDecider.decidePolicy(for:preferences:)):
(TestNavigationDecider.decidePolicy(for:)):
(WebPageTests.decidePolicyForNavigationActionFragment):
(WebPageTests.navigationPreferencesMutationDuringNavigation):
(WebPageTests.decidePolicyForNavigationResponse):

Canonical link: <a href="https://commits.webkit.org/287483@main">https://commits.webkit.org/287483@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/254baebe854a7ed1a64a4cff439cfe949e1b6a47

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/79861 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/58861 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/33259 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/84379 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/30851 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/81970 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/67929 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/7151 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/84379 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/30851 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/82929 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/52473 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/72730 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/84379 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/49813 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/26879 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/29304 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/70938 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/27354 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/85809 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/7086 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/4966 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/70682 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/7260 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/68570 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/69925 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/12852 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12346 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/7047 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/12607 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/6901 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/10413 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/8707 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Tools/TestWebKitAPI/Tests/WebKit Swift/WebPage.swift:
```diff
@@ -214,6 +244,73 @@ struct WebPageTests {
+        <script>
+        window.location.href='#fragment';
+        </script>
+        """
```
Nit - indentation (here are below as well)

whsieh commented on Tools/TestWebKitAPI/Tests/WebKit Swift/WebPage.swift:
```diff
@@ -214,6 +244,73 @@ struct WebPageTests {
+        let page = WebPage_v0(navigationDecider: decider)
+
+        let html = """
+        <script>var foo = 'bar'</script>
```
Nit - does this need to be a multiline string? Seems like it would be cleaner as a regular string literal.

rr-codes committed:
[SwiftUI] Introduce the NavigationDeciding protocol
https://bugs.webkit.org/show_bug.cgi?id=283805
rdar://140672055

Reviewed by Wenson Hsieh.

This protocol is designed to provide the subset of functionality of WKNavigationDelegate for making navigation
decisions and responding to authentication challenges.

The exception is `webView(_:authenticationChallenge:shouldAllowDeprecatedTLS:)`, which is intentionally excluded
as its functionality is effectively deprecated.

* Source/WebKit/UIProcess/API/Swift/WKNavigationDelegateAdapter.swift:
(WKNavigationDelegateAdapter.webView(_:decidePolicyFor:preferences:WKWebpagePreferences:)):
(WKNavigationDelegateAdapter.webView(_:decidePolicyFor:)):
(WKNavigationDelegateAdapter.webView(_:respondTo:URLCredential:)):
* Source/WebKit/UIProcess/API/Swift/WebPage+FrameInfo.swift: Added.
(FrameInfo.isMainFrame):
(FrameInfo.request):
(FrameInfo.securityOrigin):
(FrameInfo.wrapped):
* Source/WebKit/UIProcess/API/Swift/WebPage+NavigationDeciding.swift: Added.
(NavigationAction.source):
(NavigationAction.target):
(NavigationAction.navigationType):
(NavigationAction.request):
(NavigationAction.shouldPerformDownload):
(NavigationAction.buttonNumber):
(NavigationAction.modifierFlags):
(NavigationAction.wrapped):
(NavigationResponse.isForMainFrame):
(NavigationResponse.response):
(NavigationResponse.canShowMimeType):
(NavigationResponse.wrapped):
(NavigationDeciding.decidePolicy(for:preferences:)):
(NavigationDeciding.decidePolicy(for:)):
(NavigationDeciding.decideAuthenticationChallengeDisposition(for:URLCredential:)):
* Source/WebKit/UIProcess/API/Swift/WebPage+NavigationPreferences.swift: Added.
(preferredContentMode):
(allowsContentJavaScript):
(preferredHTTPSNavigationPolicy):
(isLockdownModeEnabled):
* Source/WebKit/UIProcess/API/Swift/WebPage.swift:
(WebPage_v0.callAsyncJavaScript(_:arguments:in:contentWorld:)):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKit Swift/Bundle+Extras.swift: Added.
(Bundle.testResources):
* Tools/TestWebKitAPI/Tests/WebKit Swift/WKWebViewSwiftOverlay.swift:
* Tools/TestWebKitAPI/Tests/WebKit Swift/WebPage.swift:
(TestNavigationDecider.preferencesMutation):
(TestNavigationDecider.decidePolicy(for:preferences:)):
(TestNavigationDecider.decidePolicy(for:)):
(WebPageTests.decidePolicyForNavigationActionFragment):
(WebPageTests.navigationPreferencesMutationDuringNavigation):
(WebPageTests.decidePolicyForNavigationResponse):

Canonical link: https://commits.webkit.org/287483@main
